data_analysis/correction_framework.py

```diff
@@ -0,0 +1,247 @@
+
+        # if the moon would have been obscured by bad weather for more then three hours, invalidate based on weather conditions
+        # use something other than python-weather?
+
+    # given a function on a float, updates all temperature, frequency, or voltage values
+    def apply_correction(self, field_to_update: str, funct):
+        if field_to_update == "TEMP":
+            for rec in self.min_records:
+                rec.temperature = funct(rec.temperature)
+        if field_to_update == "FREQ":
+            for rec in self.min_records:
+                rec.frequency = funct(rec.frequency)
+        if field_to_update == "VOLT":
+            for rec in self.min_records:
+                rec.voltage = funct(rec.voltage)
+        else:
+            print("error: field not recognized/field should not be updated")
+
+    # export night record to a new file
+    def rec_export(self, filename: str):
+        new_file = open(filename, "x")
+        if self.valid_for_use == False:
+            new_file.write("INVALID\n")
+            new_file.write(self.exclusion_reason)
+        else:
+            # night date -- 2023-01-23 06:00:00.992
+            new_file.write(self.night_date.strptime("%Y-%m-%d %H:%M:%S.%f"))
+            new_file.write("\n")
+            # morning date
+            new_file.write(self.morning_date.strptime("%Y-%m-%d %H:%M:%S.%f"))
+            new_file.write("\n")
+            # sunrise
+            new_file.write(self.sunrise.strptime("%Y-%m-%d %H:%M:%S.%f"))
+            new_file.write("\n")
+            # sunset
+            new_file.write(self.sunset.strptime("%Y-%m-%d %H:%M:%S.%f"))
+            new_file.write("\n")
+            # twilight
+            new_file.write(
+                self.astronomical_twilight.strptime("%Y-%m-%d %H:%M:%S.%f")
+            )
+            new_file.write("\n")
+            # illumination
+            new_file.write(str(self.moon_illumination))
+            new_file.write("\n")
+            # weather
+            # unimplemented
+
+            for rec in self.min_records:
+                new_file.write(
+                    rec.utc_timesunset.strptime("%Y-%m-%d %H:%M:%S.%f")
+                )
+                new_file.write(";")
+                new_file.write(rec.local_time.strptime("%Y-%m-%d %H:%M:%S.%f"))
+                new_file.write(";")
+                new_file.write(str(rec.temperature))
+                new_file.write(";")
+                new_file.write(str(rec.frequency))
+                new_file.write(";")
+                new_file.write(str(rec.voltage))
+                new_file.write(";")
+                new_file.write(str(rec.sensor_id))
+                new_file.write("\n")
+        new_file.close()
+
+    # import previously exported night record
+    def rec_import(self, filename: str):
+        # takes an already created object and updates it to imported values
+        file = open(filename)
```

Comment:
Try to use this type of file opener and you can avoid having to manage the `file.close()` function: https://www.freecodecamp.org/news/with-open-in-python-with-statement-syntax-example/#howdoesthewithstatementworkinpython

Reply:
Apply the same to the rest of the file openings as well.